src/resources/ble_esp32.cc

```diff
@@ -575,12 +578,24 @@ class ServiceContainer : public BleErrorCapableResource {
   ServiceContainer(BleResourceGroup* group, Kind kind)
       : BleErrorCapableResource(group, kind) {}
 
+  ~ServiceContainer() override {
+    while (!services_.is_empty()) {
+      auto service = services_.first();
+      group()->unregister_resource(service);
+    }
+  }
+
+  void remove_service(BleServiceResource* service) {
+    services_.unlink(service);
+  }
+
   virtual T* type() = 0;
   BleServiceResource* get_or_create_service_resource(ble_uuid_any_t uuid, uint16 start, uint16 end);
   ServiceResourceList& services() { return services_; }
 
   void clear_services() {
-    while (auto service = services_.remove_first()) {
+    while (!services_.is_empty()) {
+      auto service = services_.first();
       group()->unregister_resource(service);
```

Comment:
Maybe add a comment that explains that unregistering will remove the element from linked list?

Reply:
done.